x-pack/filebeat/input/http_endpoint/input_test.go

```diff
@@ -0,0 +1,195 @@
+				ContentType:   "application/json",
+			},
+		}},
+		events: []target{
+			{url: "http://127.0.0.1:9001/", event: `{"a":1}`},
+			{url: "http://127.0.0.1:9001/", event: `{"b":2}`},
+			{url: "http://127.0.0.1:9001/", event: `{"c":3}`},
+		},
+		want: []mapstr.M{
+			{"json": mapstr.M{"a": int64(1)}},
+			{"json": mapstr.M{"b": int64(2)}},
+			{"json": mapstr.M{"c": int64(3)}},
+		},
+	},
+	{
+		name: "distinct_ports",
+		cfgs: []*httpEndpoint{
+			{
+				addr: "127.0.0.1:9001",
+				config: config{
+					ResponseCode:  200,
+					ResponseBody:  `{"message": "success"}`,
+					ListenAddress: "127.0.0.1",
+					ListenPort:    "9001",
+					URL:           "/a/",
+					Prefix:        "json",
+					ContentType:   "application/json",
+				},
+			},
+			{
+				addr: "127.0.0.1:9002",
+				config: config{
+					ResponseCode:  200,
+					ResponseBody:  `{"message": "success"}`,
+					ListenAddress: "127.0.0.1",
+					ListenPort:    "9002",
+					URL:           "/b/",
+					Prefix:        "json",
+					ContentType:   "application/json",
+				},
+			},
+		},
+		events: []target{
+			{url: "http://127.0.0.1:9001/a/", event: `{"a":1}`},
+			{url: "http://127.0.0.1:9002/b/", event: `{"b":2}`},
+			{url: "http://127.0.0.1:9001/a/", event: `{"c":3}`},
+		},
+		want: []mapstr.M{
+			{"json": mapstr.M{"a": int64(1)}},
+			{"json": mapstr.M{"b": int64(2)}},
+			{"json": mapstr.M{"c": int64(3)}},
+		},
+	},
+	{
+		name: "shared_ports",
+		cfgs: []*httpEndpoint{
+			{
+				addr: "127.0.0.1:9001",
+				config: config{
+					ResponseCode:  200,
+					ResponseBody:  `{"message": "success"}`,
+					ListenAddress: "127.0.0.1",
+					ListenPort:    "9001",
+					URL:           "/a/",
+					Prefix:        "json",
+					ContentType:   "application/json",
+				},
+			},
+			{
+				addr: "127.0.0.1:9001",
+				config: config{
+					ResponseCode:  200,
+					ResponseBody:  `{"message": "success"}`,
+					ListenAddress: "127.0.0.1",
+					ListenPort:    "9001",
+					URL:           "/b/",
+					Prefix:        "json",
+					ContentType:   "application/json",
+				},
+			},
+		},
+		events: []target{
+			{url: "http://127.0.0.1:9001/a/", event: `{"a":1}`},
+			{url: "http://127.0.0.1:9001/b/", event: `{"b":2}`},
+			{url: "http://127.0.0.1:9001/a/", event: `{"c":3}`},
+		},
+		want: []mapstr.M{
+			{"json": mapstr.M{"a": int64(1)}},
+			{"json": mapstr.M{"b": int64(2)}},
+			{"json": mapstr.M{"c": int64(3)}},
+		},
+	},
+}
+
+type target struct {
+	url   string
+	event string
+}
+
+func TestServerPool(t *testing.T) {
+	for _, test := range serverPoolTests {
+		t.Run(test.name, func(t *testing.T) {
+			servers := pool{servers: make(map[string]*server)}
+
+			var pub publisher
+			ctx, cancel := newCtx("server_pool_test", test.name)
+			var wg sync.WaitGroup
+			for _, cfg := range test.cfgs {
+				cfg := cfg
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					_ = servers.serve(ctx, cfg, &pub) // Always returns a non-nil error.
+				}()
+			}
+			time.Sleep(time.Second)
```

Comment:
I'm not particularly proud of this, but I don't see another way to ensure that we have the servers up without doing nasty things inside `pool.serve`.